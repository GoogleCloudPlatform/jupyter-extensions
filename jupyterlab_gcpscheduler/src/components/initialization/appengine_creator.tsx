/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { css, LearnMoreLink } from 'gcp_jupyterlab_shared';
import { stylesheet } from 'typestyle';

interface Props {
  projectId: string;
}

const SUPPORTED_REGIONS_LINK =
  'https://cloud.google.com/scheduler/docs/#supported_regions';
const APPENGINE_CREATE_LINK =
  'https://console.cloud.google.com/appengine/start';
const CLOUD_SHELL_LINK = 'https://ssh.cloud.google.com/cloudshell/editor';

const localStyles = stylesheet({
  code: {
    backgroundColor: '#f7f7f7',
    display: 'block',
    fontFamily: '"Roboto Mono", monospace',
    margin: '16px 0',
    padding: '8px',
  },
  li: {
    margin: '8px',
  },
});

/** Responsible for creating an AppEngine Project. */
export function AppEngineCreator(props: Props) {
  return (
    <div className={css.column}>
      <p className={css.paragraph}>
        In order to use Cloud Scheduler, your project must contain an App Engine
        app that is located in one of the supported regions
        <LearnMoreLink href={SUPPORTED_REGIONS_LINK} />. This is a one-time
        setup that must be completed by a Project Owner. There are two ways to
        create the App Engine app:
      </p>
      <ol>
        <li className={localStyles.li}>
          Use the
          <LearnMoreLink
            href={`${APPENGINE_CREATE_LINK}?project=${props.projectId}`}
            text="Google Cloud Platform Console UI"
          />
        </li>
        <li className={localStyles.li}>
          Run the following command in a terminal or in
          <LearnMoreLink href={CLOUD_SHELL_LINK} text="Google Cloud Shell" />:
          <code className={localStyles.code}>
            gcloud --project {props.projectId} app create
          </code>
        </li>
      </ol>
    </div>
  );
}
