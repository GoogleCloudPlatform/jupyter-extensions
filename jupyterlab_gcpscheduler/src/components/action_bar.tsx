/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import * as React from 'react';
import { style } from 'typestyle';

import { css } from 'gcp_jupyterlab_shared';
import { OnDialogClose } from './dialog';

interface Props {
  children?: React.ReactNode;
  closeLabel?: string;
  onDialogClose: OnDialogClose;
}

const actionBar = style({
  paddingTop: '16px',
  paddingRight: '2px',
  $nest: {
    '&>*': {
      marginLeft: '16px',
    },
  },
  ...csstips.horizontal,
  ...csstips.endJustified,
});

/** Funtional Component for defining an action bar with buttons. */
export function ActionBar(props: Props) {
  return (
    <div className={actionBar}>
      <button
        type="button"
        className={css.button}
        onClick={props.onDialogClose}
      >
        {props.closeLabel || 'Close'}
      </button>
      {props.children}
    </div>
  );
}
