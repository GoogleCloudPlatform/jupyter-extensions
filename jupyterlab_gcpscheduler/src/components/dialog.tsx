/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ISettingRegistry } from '@jupyterlab/coreutils';
import { INotebookModel } from '@jupyterlab/notebook';
import { Dialog } from '@material-ui/core';
import * as csstips from 'csstips';
import {
  BASE_FONT,
  COLORS,
  css,
  IconButtonMenu,
  MenuCloseHandler,
  SmallMenuItem,
  Message,
} from 'gcp_jupyterlab_shared';
import * as React from 'react';
import { stylesheet } from 'typestyle';

import { SCHEDULER_LINK, BUCKET_NAME_SUFFIX } from '../data';
import { GcpService } from '../service/gcp';
import {
  GetPermissionsResponse,
  ProjectStateService,
  ProjectState,
} from '../service/project_state';
import { Initializer } from './initialization/initializer';
import { SchedulerForm } from './scheduler_form';
import { ActionBar } from './action_bar';

import {
  ClientTransportService,
  ServerProxyTransportService,
} from 'gcp_jupyterlab_shared';

/** Information provided to the GcpSchedulerWidget */
export interface LaunchSchedulerRequest {
  timestamp: number;
  notebookName: string;
  notebook: INotebookModel;
}

export interface JobSubmittedMessage {
  message: string;
  link: string;
}

/** Definition for a function that closes the SchedulerDialog. */
export type OnDialogClose = () => void;

/** Callback function that accepts an initialized ProjectState */
export type OnInitialized = (projectState: ProjectState) => void;

/** Extension settings. */
export interface GcpSettings {
  projectId: string;
  gcsBucket: string;
  schedulerRegion: string;
  jobRegion?: string;
  scaleTier?: string;
  masterType?: string;
  acceleratorType?: string;
  acceleratorCount?: string;
  containerImage?: string;
  oAuthClientId?: string;
}

interface Props {
  projectStateService: ProjectStateService;
  gcpService: GcpService;
  request: LaunchSchedulerRequest;
  settings: ISettingRegistry.ISettings;
}

interface State {
  canSchedule: boolean;
  dialogClosedByUser: boolean;
  gcpSettings?: GcpSettings;
  permissions?: GetPermissionsResponse;
  jobSubmittedMessage?: JobSubmittedMessage;
}

const localStyles = stylesheet({
  header: {
    ...BASE_FONT,
    fontWeight: 500,
    fontSize: '15px',
    margin: '16px 16px 0 16px',
    ...csstips.horizontal,
    ...csstips.center,
  },
  title: {
    ...csstips.flex,
  },
  main: {
    backgroundColor: COLORS.white,
    color: COLORS.base,
    padding: '16px',
    width: '480px',
    ...BASE_FONT,
    ...csstips.vertical,
  },
});

const PYTHON2 = 'python2';
const PYTHON2_WARNING =
  'Python 2 Notebooks are not supported. Please upgrade your Notebook to use Python 3';

/**
 * Dialog wrapping the GCP scheduler UI.
 */
export class SchedulerDialog extends React.Component<Props, State> {
  private getPermissionsPromise?: Promise<void>;

  constructor(props: Props) {
    super(props);

    this.state = {
      dialogClosedByUser: false,
      canSchedule: false,
    };
    this._settingsChanged = this._settingsChanged.bind(this);
    this._onDialogClose = this._onDialogClose.bind(this);
    this._onInitialized = this._onInitialized.bind(this);
  }

  /** Establishes the binding for Settings Signal and invokes the handler. */
  componentDidMount() {
    this.props.settings.changed.connect(this._settingsChanged);
    this._settingsChanged(this.props.settings);
  }

  componentWillUnmount() {
    this.props.settings.changed.disconnect(this._settingsChanged);
  }

  /**
   * Set the dialog to be open since for any new request with a valid Notebook.
   */
  componentDidUpdate(prevProps: Props) {
    if (
      prevProps.request !== this.props.request &&
      !!this.props.request.notebook
    ) {
      this.setState({ dialogClosedByUser: false });
    }
    if (
      this._isOpen() &&
      !this.state.permissions &&
      !this.getPermissionsPromise
    ) {
      this.getPermissionsPromise = this._checkPermissions();
    }
  }

  render() {
    const { gcpSettings } = this.state;
    const projectId = gcpSettings ? gcpSettings.projectId : '';
    return (
      <Dialog open={this._isOpen()}>
        <header className={localStyles.header}>
          <span className={localStyles.title}>Schedule a notebook run</span>
          <IconButtonMenu
            menuItems={menuCloseHandler => [
              <SmallMenuItem key="viewAllJobs">
                <a
                  href={`${SCHEDULER_LINK}?project=${projectId}`}
                  target="_blank"
                  onClick={menuCloseHandler}
                >
                  View all scheduled jobs
                </a>
              </SmallMenuItem>,
              <SmallMenuItem
                key="reset"
                onClick={() => this._onResetSettings(menuCloseHandler)}
              >
                Reset configuration
              </SmallMenuItem>,
            ]}
          ></IconButtonMenu>
        </header>
        <main className={localStyles.main}>{this._getDialogContent()}</main>
      </Dialog>
    );
  }

  private _getDialogContent(): JSX.Element {
    const {
      canSchedule,
      gcpSettings,
      permissions,
      jobSubmittedMessage,
    } = this.state;
    const { gcpService, projectStateService, request } = this.props;
    const hasNotebook = !!(request && request.notebook);
    if (jobSubmittedMessage) {
      return (
        <div className={css.column}>
          <Message text={jobSubmittedMessage.message} />
          <ActionBar onDialogClose={this._onDialogClose}>
            <a
              href={jobSubmittedMessage.link}
              target="_blank"
              className={css.button}
            >
              View Job
            </a>
          </ActionBar>
        </div>
      );
    } else if (
      hasNotebook &&
      request.notebook.defaultKernelName.toLowerCase().replace(' ', '') ===
        PYTHON2
    ) {
      // For now, only exclude Python 2 kernels
      return (
        <div className={css.column}>
          <Message asError={true} text={PYTHON2_WARNING} />
          <ActionBar onDialogClose={this._onDialogClose} />
        </div>
      );
    } else if (!canSchedule) {
      return (
        <Initializer
          projectStateService={projectStateService}
          onDialogClose={this._onDialogClose}
          onInitialized={this._onInitialized}
        />
      );
    } else if (canSchedule && hasNotebook) {
      return permissions ? (
        <SchedulerForm
          gcpService={gcpService}
          gcpSettings={gcpSettings}
          notebookName={request.notebookName}
          notebook={request.notebook}
          permissions={permissions}
          onDialogClose={this._onDialogClose}
          settings={this.props.settings}
        />
      ) : (
        <div className={css.column}>
          <Message asActivity={true} text="Checking IAM Permissions" />
          <ActionBar onDialogClose={this._onDialogClose} />
        </div>
      );
    }
    return null;
  }

  // Casts to GcpSettings shape from JSONObject
  private _settingsChanged(newSettings: ISettingRegistry.ISettings) {
    const settings = (newSettings.composite as unknown) as GcpSettings;
    const canSchedule = !!(settings.projectId && settings.gcsBucket);
    if (settings.projectId) {
      console.info(`Using ${settings.projectId} for GCP API calls`);
      this.props.projectStateService.projectId = settings.projectId;
    }
    if (settings.oAuthClientId) {
      console.info('Using end-user authentication for GCP API calls');
      this.props.gcpService.transportService = new ClientTransportService(
        settings.oAuthClientId
      );
    } else if (this.state.gcpSettings && this.state.gcpSettings.oAuthClientId) {
      console.info('Using machine authentication for GCP API calls');
      this.props.gcpService.transportService = new ServerProxyTransportService();
    }

    this.setState({ gcpSettings: settings, canSchedule });
  }

  // Callback when Initializer has a valid project state
  private _onInitialized(projectState: ProjectState) {
    const { settings } = this.props;
    const gcpSettings = (settings.composite as unknown) as GcpSettings;
    if (gcpSettings.projectId !== projectState.projectId) {
      settings.set('projectId', projectState.projectId);
    }
    if (!gcpSettings.gcsBucket && projectState.hasGcsBucket) {
      settings.set(
        'gcsBucket',
        `gs://${projectState.projectId}${BUCKET_NAME_SUFFIX}`
      );
    }
    if (!gcpSettings.schedulerRegion && projectState.canSubmitScheduledJobs) {
      settings.set('schedulerRegion', projectState.schedulerRegion);
    }
  }

  private _onDialogClose() {
    this.setState({ dialogClosedByUser: true });
  }

  private _onResetSettings(closeHandler: MenuCloseHandler) {
    this.props.settings.save('{}');
    closeHandler();
  }

  private _isOpen() {
    const hasNotebook = !!(this.props.request && this.props.request.notebook);
    return hasNotebook && !this.state.dialogClosedByUser;
  }

  private async _checkPermissions() {
    const permissions = await this.props.projectStateService.getPermissions();
    this.setState({ permissions });
  }
}
