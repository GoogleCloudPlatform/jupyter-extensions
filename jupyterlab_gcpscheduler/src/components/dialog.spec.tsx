/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ISettingRegistry } from '@jupyterlab/coreutils';
import { NBTestUtils } from '@jupyterlab/testutils';
import { Dialog } from '@material-ui/core';
import { shallow } from 'enzyme';
import { IconButtonMenu } from 'gcp_jupyterlab_shared';
import * as React from 'react';

import { GcpService } from '../service/gcp';
import { LaunchSchedulerRequest, SchedulerDialog, GcpSettings } from './dialog';
import { INotebookModel } from '@jupyterlab/notebook';
import { TEST_PROJECT, getProjectState } from '../test_helpers';
import {
  ClientTransportService,
  ServerProxyTransportService,
  TransportService,
} from 'gcp_jupyterlab_shared';
import { ProjectStateService, ProjectState } from '../service/project_state';
import { ActionBar } from './action_bar';
import { Initializer } from './initialization/initializer';
import { BUCKET_NAME_SUFFIX } from '../data';

describe('SchedulerDialog', () => {
  const getPermissionsPromise = Promise.resolve({
    toInitialize: [],
    toExecute: [],
    toSchedule: [],
  });
  const mockGetProjectState = jest.fn();
  const mockSetProjectId = jest.fn();
  const mockSetTransportService = jest.fn();
  const mockSettingsChangedConnect = jest.fn();
  const mockSettingsSet = jest.fn();
  const mockGetPermissions = jest.fn();
  const mockProjectStateService = ({
    set projectId(projectId: string) {
      mockSetProjectId(projectId);
    },
    getProjectState: mockGetProjectState,
    getPermissions: mockGetPermissions,
  } as unknown) as ProjectStateService;
  const mockGcpService = {
    set transportService(transportService: TransportService) {
      mockSetTransportService(transportService);
    },
  } as GcpService;
  const fakeNotebook = NBTestUtils.createNotebook();
  NBTestUtils.populateNotebook(fakeNotebook);
  const launchSchedulerRequest: LaunchSchedulerRequest = {
    timestamp: Date.now(),
    notebook: null,
    notebookName: null,
  };

  function getSettings(gcpSettings: Partial<GcpSettings>) {
    return ({
      changed: {
        connect: mockSettingsChangedConnect,
        disconnect: jest.fn(),
      },
      composite: gcpSettings,
      save: jest.fn(),
      set: mockSettingsSet,
    } as unknown) as ISettingRegistry.ISettings;
  }

  beforeEach(() => {
    jest.resetAllMocks();
    jest.useFakeTimers();
    mockGetPermissions.mockReturnValue(getPermissionsPromise);
    launchSchedulerRequest.notebookName = null;
    launchSchedulerRequest.notebook = null;
  });

  it('Renders closed Dialog without Notebook', async () => {
    const settings = getSettings({});
    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(settings.changed.connect).toHaveBeenCalled();
    expect(dialog).toMatchSnapshot('Dialog Closed');
  });

  it('Renders with Python2 warning', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'p2-nb.ipynb';
    launchSchedulerRequest.notebook = {
      defaultKernelName: 'Python 2',
    } as INotebookModel;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(settings.changed.connect).toHaveBeenCalled();
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    expect(dialog).toMatchSnapshot('Python2Warning');
  });

  it('Renders with Checking IAM Message', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(mockGetPermissions).toHaveBeenCalledTimes(1);
    expect(settings.changed.connect).toHaveBeenCalled();
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    expect(dialog).toMatchSnapshot('IamMessage');
  });

  it('Renders with SchedulerForm', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    await getPermissionsPromise;
    expect(mockGetPermissions).toHaveBeenCalledTimes(1);
    expect(settings.changed.connect).toHaveBeenCalled();
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    expect(dialog).toMatchSnapshot('SchedulerForm');
  });

  it('Renders with Initializer', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    expect(dialog).toMatchSnapshot('Initializer');
  });

  it('Closes dialog when clicked', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(dialog.state('dialogClosedByUser')).toBe(false);
    dialog
      .find(ActionBar)
      .dive()
      .find('button')
      .simulate('click');
    expect(dialog.state('dialogClosedByUser')).toBe(true);
  });

  it('Updates settings from Initializer', async () => {
    const settings = getSettings({});
    const projectState: ProjectState = {
      ...getProjectState(),
      canSubmitScheduledJobs: true,
      hasGcsBucket: true,
      schedulerRegion: 'us-central1',
    };
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    dialog.find(Initializer).prop('onInitialized')(projectState);
    expect(mockSettingsSet).toHaveBeenCalledWith(
      'projectId',
      projectState.projectId
    );
    expect(mockSettingsSet).toHaveBeenCalledWith(
      'gcsBucket',
      `gs://${projectState.projectId}${BUCKET_NAME_SUFFIX}`
    );
    expect(mockSettingsSet).toHaveBeenCalledWith(
      'schedulerRegion',
      projectState.schedulerRegion
    );
  });

  it('Clears settings when Reset is clicked', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );

    dialog
      .find(IconButtonMenu)
      .dive()
      .findWhere(w => w.text() === 'Reset configuration')
      .first()
      .simulate('click');
    expect(settings.save).toHaveBeenCalledWith('{}');
  });

  it('Reopens a closed dialog when request prop changes', async () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      gcsBucket: 'gs://test-project/notebooks',
      schedulerRegion: 'us-east1',
    });
    launchSchedulerRequest.notebookName = 'Foo.ipynb';
    launchSchedulerRequest.notebook = fakeNotebook.model;

    const dialog = shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(dialog.find(Dialog).prop('open')).toBe(true);
    dialog.setState({ dialogClosedByUser: true });
    expect(dialog.find(Dialog).prop('open')).toBe(false);

    const newRequest: LaunchSchedulerRequest = {
      timestamp: Date.now(),
      notebookName: 'A different notebook.ipynb',
      notebook: launchSchedulerRequest.notebook,
    };
    dialog.setProps({ request: newRequest });
    expect(dialog.find(Dialog).prop('open')).toBe(true);
  });

  it('Uses ClientTransportService when OAuth Client ID is set', () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      oAuthClientId: 'test.apps.googleusercontent.com',
    });

    shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    const setTransportServiceCalls = mockSetTransportService.mock.calls;
    expect(setTransportServiceCalls.length).toBe(1);
    expect(setTransportServiceCalls[0][0]).toBeInstanceOf(
      ClientTransportService
    );
  });

  it('Uses ServerProxyTransportService when OAuth Client ID is removed', () => {
    const settings = getSettings({
      projectId: TEST_PROJECT,
      oAuthClientId: 'test.apps.googleusercontent.com',
    });

    mockSettingsChangedConnect.mockImplementation(handler => {
      setTimeout(() => {
        handler(getSettings({ projectId: TEST_PROJECT }));
      }, 1);
    });

    shallow(
      <SchedulerDialog
        projectStateService={mockProjectStateService}
        gcpService={mockGcpService}
        request={launchSchedulerRequest}
        settings={settings}
      />
    );
    const setTransportServiceCalls = mockSetTransportService.mock.calls;
    expect(mockSetProjectId).toHaveBeenCalledWith(TEST_PROJECT);
    expect(setTransportServiceCalls.length).toBe(1);
    expect(setTransportServiceCalls[0][0]).toBeInstanceOf(
      ClientTransportService
    );
    jest.advanceTimersByTime(1);
    expect(setTransportServiceCalls.length).toBe(2);
    expect(setTransportServiceCalls[1][0]).toBeInstanceOf(
      ServerProxyTransportService
    );
  });
});
