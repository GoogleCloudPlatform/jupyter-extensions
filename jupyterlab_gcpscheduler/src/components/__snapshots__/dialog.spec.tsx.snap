// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchedulerDialog Renders closed Dialog without Notebook: Dialog Closed 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={false}
>
  <header
    className="header_fst2s1z"
  >
    <span
      className="title_fiw0lep"
    >
      Schedule a notebook run
    </span>
    <IconButtonMenu
      menuItems={[Function]}
    />
  </header>
  <main
    className="main_f2ra5xi"
  >
    <Initializer
      onDialogClose={[Function]}
      onInitialized={[Function]}
      projectStateService={
        Object {
          "getPermissions": [MockFunction],
          "getProjectState": [MockFunction],
          "projectId": undefined,
        }
      }
    />
  </main>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`SchedulerDialog Renders with Checking IAM Message: IamMessage 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <header
    className="header_fst2s1z"
  >
    <span
      className="title_fiw0lep"
    >
      Schedule a notebook run
    </span>
    <IconButtonMenu
      menuItems={[Function]}
    />
  </header>
  <main
    className="main_f2ra5xi"
  >
    <div
      className="column_fghohlb"
    >
      <Message
        asActivity={true}
        text="Checking IAM Permissions"
      />
      <ActionBar
        onDialogClose={[Function]}
      />
    </div>
  </main>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`SchedulerDialog Renders with Initializer: Initializer 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <header
    className="header_fst2s1z"
  >
    <span
      className="title_fiw0lep"
    >
      Schedule a notebook run
    </span>
    <IconButtonMenu
      menuItems={[Function]}
    />
  </header>
  <main
    className="main_f2ra5xi"
  >
    <Initializer
      onDialogClose={[Function]}
      onInitialized={[Function]}
      projectStateService={
        Object {
          "getPermissions": [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "getProjectState": [MockFunction],
          "projectId": undefined,
        }
      }
    />
  </main>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`SchedulerDialog Renders with Python2 warning: Python2Warning 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <header
    className="header_fst2s1z"
  >
    <span
      className="title_fiw0lep"
    >
      Schedule a notebook run
    </span>
    <IconButtonMenu
      menuItems={[Function]}
    />
  </header>
  <main
    className="main_f2ra5xi"
  >
    <div
      className="column_fghohlb"
    >
      <Message
        asError={true}
        text="Python 2 Notebooks are not supported. Please upgrade your Notebook to use Python 3"
      />
      <ActionBar
        onDialogClose={[Function]}
      />
    </div>
  </main>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`SchedulerDialog Renders with SchedulerForm: SchedulerForm 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <header
    className="header_fst2s1z"
  >
    <span
      className="title_fiw0lep"
    >
      Schedule a notebook run
    </span>
    <IconButtonMenu
      menuItems={[Function]}
    />
  </header>
  <main
    className="main_f2ra5xi"
  >
    <WithFormik(InnerSchedulerForm)
      gcpService={
        Object {
          "transportService": undefined,
        }
      }
      gcpSettings={
        Object {
          "gcsBucket": "gs://test-project/notebooks",
          "projectId": "test-project",
          "schedulerRegion": "us-east1",
        }
      }
      notebook={
        Object {
          "cells": Array [
            Object {
              "cell_type": "code",
              "execution_count": 1,
              "metadata": Object {
                "tags": Array [],
              },
              "outputs": Array [
                Object {
                  "name": "stdout",
                  "output_type": "stream",
                  "text": "hello world

0

1

2
",
                },
                Object {
                  "name": "stderr",
                  "output_type": "stream",
                  "text": "output to stderr
",
                },
                Object {
                  "name": "stdout",
                  "output_type": "stream",
                  "text": "some more stdout text
",
                },
              ],
              "source": "import sys
sys.stdout.write('hello world\\\\n')
sys.stdout.flush()
for i in range(3):
    sys.stdout.write('%s\\\\n' % i)
    sys.stdout.flush()
sys.stderr.write('output to stderr\\\\n')
sys.stderr.flush()
sys.stdout.write('some more stdout text\\\\n')
sys.stdout.flush()",
            },
            Object {
              "cell_type": "markdown",
              "metadata": Object {
                "tags": Array [],
              },
              "source": "# Markdown Cell

$ e^{ \\\\pm i\\\\theta } = \\\\cos \\\\theta \\\\pm i\\\\sin \\\\theta + \\\\beta $

*It* **really** is!",
            },
            Object {
              "cell_type": "code",
              "execution_count": 2,
              "metadata": Object {
                "tags": Array [],
              },
              "outputs": Array [
                Object {
                  "ename": "SyntaxError",
                  "evalue": "invalid syntax (<ipython-input-2-6c5185427360>, line 1)",
                  "output_type": "error",
                  "traceback": Array [
                    "[0;36m  File [0;32m\\"<ipython-input-2-6c5185427360>\\"[0;36m, line [0;32m1[0m
[0;31m    this is a syntax error[0m
[0m                   ^[0m
[0;31mSyntaxError[0m[0;31m:[0m invalid syntax
",
                  ],
                },
              ],
              "source": "this is a syntax error",
            },
            Object {
              "cell_type": "code",
              "execution_count": null,
              "metadata": Object {
                "tags": Array [],
              },
              "outputs": Array [],
              "source": "print('test')",
            },
            Object {
              "cell_type": "code",
              "execution_count": 4,
              "metadata": Object {
                "tags": Array [],
              },
              "outputs": Array [
                Object {
                  "data": Object {
                    "text/latex": Array [
                      "The mass-energy equivalence is described by the famous equation
",
                      " 
",
                      "$$E=mc^2$$
",
                      " 
",
                      "discovered in 1905 by Albert Einstein. 
",
                      "In natural units ($c$ = 1), the formula expresses the identity
",
                      " 
",
                      "\\\\begin{equation}
",
                      "E=m
",
                      "\\\\end{equation}",
                    ],
                    "text/plain": Array [
                      "<IPython.core.display.Latex object>",
                    ],
                  },
                  "execution_count": 4,
                  "metadata": Object {},
                  "output_type": "execute_result",
                },
              ],
              "source": "from IPython.display import Latex
Latex('''The mass-energy equivalence is described by the famous equation
 
$$E=mc^2$$
 
discovered in 1905 by Albert Einstein. 
In natural units ($c$ = 1), the formula expresses the identity
 
\\\\\\\\begin{equation}
E=m
\\\\\\\\end{equation}''')",
            },
            Object {
              "cell_type": "code",
              "execution_count": null,
              "metadata": Object {
                "collapsed": true,
                "jupyter": Object {
                  "outputs_hidden": true,
                },
              },
              "outputs": Array [],
              "source": "",
            },
          ],
          "metadata": Object {
            "anaconda-cloud": Object {},
            "kernelspec": Object {
              "display_name": "Python [default]",
              "language": "python",
              "name": "python3",
            },
            "language_info": Object {
              "codemirror_mode": Object {
                "name": "ipython",
                "version": 3,
              },
              "file_extension": ".py",
              "mimetype": "text/x-python",
              "name": "python",
              "nbconvert_exporter": "python",
              "pygments_lexer": "ipython3",
              "version": "3.5.2",
            },
          },
          "nbformat": 4,
          "nbformat_minor": 4,
        }
      }
      notebookName="Foo.ipynb"
      onDialogClose={[Function]}
      permissions={
        Object {
          "toExecute": Array [],
          "toInitialize": Array [],
          "toSchedule": Array [],
        }
      }
      settings={
        Object {
          "changed": Object {
            "connect": [MockFunction] {
              "calls": Array [
                Array [
                  [Function],
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": undefined,
                },
              ],
            },
            "disconnect": [MockFunction],
          },
          "composite": Object {
            "gcsBucket": "gs://test-project/notebooks",
            "projectId": "test-project",
            "schedulerRegion": "us-east1",
          },
          "save": [MockFunction],
          "set": [MockFunction],
        }
      }
    />
  </main>
</WithStyles(ForwardRef(Dialog))>
`;
